src/hooks/useSearch.js

```diff
@@ -0,0 +1,166 @@
+import { uniqBy } from './usePagination';
+
+const searchInputDebounceMs = 500;
+
+/**
+ * useSearch hook enables a user to perform search functionality from their WordPress site
+ * with proper debouncing of the search input, and pagination via the `loadMore` function.
+ *
+ * @returns {{searchQuery: string, setSearchQuery: (newValue) => void, searchResults: object[] | null, loadMore: () => void, isLoading: boolean, pageInfo: object;}} Result object
+ */
+export default function useSearch() {
+  const [searchQuery, setSearchQuery] = useState('');
+  const [debouncedSearchQuery] = useDebounce(
+    searchQuery,
+    searchInputDebounceMs
+  );
+  const [searchResults, setSearchResults] = useState(null);
+  const [pageInfo, setPageInfo] = useState(null);
+  const [isLoading, setIsLoading] = useState(false);
+
+  /**
+   * Fetch results based on the search query and cursor if we are paginating.
+   * @param {string} searchQuery The user inputted search query
+   * @param {string | undefined} endCursor The end cursor if we are paginating
+   * @returns
+   */
+  async function fetchResults(searchQuery, endCursor = undefined) {
+    /**
+     * Do a prepass request for the search result metadata so we can make a
+     * proper subsequent request for the title and content based on the content
+     * type.
+     *
+     * typically this can be done with the ... on NodeWithTitle interface, but there is
+     * currently a bug.
+     *
+     * @see https://gqty.dev/docs/client/helper-functions#prepass
+     * @see https://github.com/gqty-dev/gqty/issues/733
+     */
+    const metadata = await client.client.resolved(() => {
+      const { nodes, pageInfo } = client.client.query.contentNodes({
+        first: appConfig?.postsPerPage,
+        after: endCursor,
+        where: { search: searchQuery },
+      });
+
+      prepass(nodes, 'databaseId', 'id', 'uri', 'date', '__typename');
+
+      return { nodes, pageInfo };
+    });
+
+    /**
+     * Make a prepass request for the title and excerpt from the previously
+     * fetched metadata.
+     */
+    const metadataWithContent = await client.client.resolved(() => {
+      metadata?.nodes?.map((node) =>
+        prepass(
+          node,
+          `$on.${node?.__typename}.title`,
+          `$on.${node?.__typename}.excerpt`
+        )
+      );
+
+      return {
+        nodes: getArrayFields(
+          metadata?.nodes,
+          'databaseId',
+          'id',
+          'uri',
+          'date',
+          '__typename',
+          '$on'
+        ),
+        pageInfo: getFields(metadata?.pageInfo),
+      };
+    });
+
+    return metadataWithContent;
+  }
+
+  /**
+   * Fetch initial results. This can happen either upon first search. Or after
+   * a search query has been deleted and the user types a new search query.
+   */
+  const fetchInitialResults = useCallback(async () => {
+    setIsLoading(true);
+
+    clearResults();
+
+    const res = await fetchResults(debouncedSearchQuery);
```

Comment:
We have to be careful here how we handle errors. If this request fails then there is no fallback.

Reply:
Good call. For now I've added a `try`/`catch` block that when an error is thrown, will clear the UI and present a message. In the future, we could use some kind of fetch retry if necessary.